truffle/src/main/java/org/jruby/truffle/nodes/coerce/ToAryNode.java

```diff
@@ -0,0 +1,59 @@
+package org.jruby.truffle.nodes.coerce;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.NodeChild;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.source.SourceSection;
+import org.jruby.truffle.nodes.RubyNode;
+import org.jruby.truffle.nodes.dispatch.CallDispatchHeadNode;
+import org.jruby.truffle.nodes.dispatch.DispatchHeadNodeFactory;
+import org.jruby.truffle.runtime.RubyContext;
+import org.jruby.truffle.runtime.control.RaiseException;
+import org.jruby.truffle.runtime.core.RubyArray;
+
+
+@NodeChild(value = "child", type = RubyNode.class)
+public abstract class ToAryNode extends RubyNode {
+
+    @Child private CallDispatchHeadNode toAryNode;
+
+    public ToAryNode(RubyContext context, SourceSection sourceSection) {
+        super(context, sourceSection);
+    }
+
+    public ToAryNode(ToAryNode prev) {
+        super(prev);
+    }
+
+    @Specialization
+    public Object coerceRubyArray(RubyArray rubyArray) {
+        return rubyArray;
+    }
+
+    @Specialization( guards = "!isRubyArray")
+    public Object coerceObject(VirtualFrame frame, Object object) {
+        notDesignedForCompilation();
+
+        if (toAryNode == null) {
+            CompilerDirectives.transferToInterpreter();
+            toAryNode = insert(DispatchHeadNodeFactory.createMethodCall(getContext()));
+        }
+
+        final Object coerced;
+
+        try {
+            coerced = toAryNode.call(frame, object, "to_ary", null);
+        } catch (RaiseException e) {
+            if (e.getRubyException().getLogicalClass() == getContext().getCoreLibrary().getNoMethodErrorClass()) {
+                CompilerDirectives.transferToInterpreter();
+
+                throw new RaiseException(
+                    getContext().getCoreLibrary().typeErrorNoImplicitConversion(object, "Array", this));
+            } else {
+                throw e;
+            }
+        }
+        return coerced;
```

Comment:
If you look at the other coercion nodes, they check the value of the return type.  I think you need to do that here as well, since someone could define `to_ary` to return something other than a `RubyArray`.  Here are the two cases I saw with `Array#replace`:

```
$ ruby -e 'class X; end; [].replace(X.new)'
-e:1:in `replace': no implicit conversion of X into Array (TypeError)
	from -e:1:in `<main>'
```
That's the missing `to_ary` implementation, which you've addressed in the code above.  But there's also:

```
$ ruby -e 'class X; def to_ary; :hello; end; end; [].replace(X.new)'
-e:1:in `replace': can't convert X to Array (X#to_ary gives Symbol) (TypeError)
	from -e:1:in `<main>'
```

That's the case you're missing.